text/0096-rawkv-stale-read.md

```diff
@@ -0,0 +1,97 @@
+
+    <p align="center">
+        <img src="../media/lease.png" width="450px"/>
+    </p>
+
+3. Stale read for TxnKV only.
+
+    Stale read on TxnKV allows to read outdated data from the replica other than the leader but still keeping snapshot isolation provided by TiKV transaction. Because the client tells the follower the timestamp of the stale snapshot it wants to read on (rather than read-the-latest), the follower can check if the data versions that are fresh enough via the `saft_ts` (the minimum timestamp of the on-the-fly transaction). Any read transaction with a smaller timestamp than `safe_ts` could be preserved the snapshot isolation.
+
+    <p align="center">
+        <img src="../media/txnkv-stale-read.svg" width="450px"/>
+    </p>
+
+This RFC proposes stale read on RawKV which allows reading stale data from the not-leader replica with a lossy bound. Unlike the stale read on TxnKV, the stale read on RawKV breaks its original linear consistency guarantee -- downgrades to eventually consistency. TiKV only guarantees that the stale data read from followers are written before (read committed).
+
+The follower will read the local storage with a `read_ts` that reuses the mechanism from the stale read of TxnKV. Surprisingly, the leader can also stale read, and in this case, the leader can read locally without a lease.
+
+Note that the freshness of data may regress from the point of view of the client since the client may choose different followers to read from time to time.
+
+<p align="center">
+    <img src="../media/rawkv-stale-read.svg" width="450px"/>
+</p>
+
+
+## Detailed Design
+
+This RFC proposes stale read on RawKV which allows reading stale data from the not-leader replica with a lossy time-bound. Unlike the stale read on TxnKV, the stale read on RawKV breaks its original linear consistency guarantee -- downgrades to eventually consistency. TiKV only guarantees that the stale data read from followers are written before.
+
+### TiKV
+
+ While trying to read data, clients should specify a timestamp which attachs to the request header as `read_ts`, typically a timestamp few seconds ago. The replica should read the local storage with the `read_ts` and reuses the mechanism from the stale read of TxnKV. This requires the replica to check the `read_ts` against the `safe_ts` which is advaneced by `CheckLeader` message from the store of the leader (for follower) or `resolve-ts` worker (for leader). As long as the `safe_ts` is no less than `read_ts`, the replica is allowed to read the key from local storage. Notice that there is no lock for the RawKV regions, thus the `resolve-ts` worker advanced the `safe_ts` by requesting the TSO for the latest timestamp.
```

Comment:
1. What's the meaning of `safe_ts` for RawKV ? Suggest to give a definition, e.g, "the minimum timestamp of the on-the-fly RawKV writes".

2. If the definition depends on "timestamp" of RawKV writes, this feature depends on the timestamp introduced by API V2, is that right ?

3. The mechanism to get the "minimal timestamp" of the on-the-fly writes between Txn & Raw would be quite different. Although there is no locks, Raw writes would still be "on-the-fly" during Raft procedure.

4. RawKV CDC faces the a very similar problem to track "on-the-fly" for [`resolved-ts`](https://github.com/pingyu/tikv-rfcs/blob/rawkv-cdc/text/0083-rawkv-cross-cluster-replication.md#23-resolved-timestamp). I think we can reuse it for stale read. Please refer to #86 .